modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,862 @@
+      return result;
+    }
+
+    /*
+      Compare wrapper
+    */
+    pragma "no doc"
+    proc const _compare(x: eltType, y: eltType) {
+      return chpl_compare(x, y, comparator);
+    }
+
+    pragma "no doc"
+    proc ref _insert(ref node: nodeType, element: eltType, parent: nodeType): bool {
+      if node == nil {
+        node = new nodeType(element, _random(), 1, parent);
+        return true;
+      }
+      var cmp = _compare(element, node!.element);
+      if cmp == 0 then return false;
+      else {
+        var result = false;
+        var pos: int = cmp > 0;
+        result = _insert(node!.children[pos], element, node);
+        if node!.children[pos]!.rank > node!.rank {
+          _rotate(node, pos);
+        }
+        node!.update();
+        return result;
+      }
+    }
+
+    pragma "no doc"
+    proc ref _remove(ref node: nodeType, const ref x: eltType): bool {
+      if node == nil then return false;
+      var cmp = _compare(x, node!.element);
+      if cmp == 0 {
+        var children = node!.children;
+        if children[0] == nil && children[1] == nil {
+          // Leaf node, safely removed
+          delete node;
+          node = nil;
+          return true;
+        }
+
+        // Choose the child with the greater rank
+        var childPos = 0;
+        if children[0] == nil {
+          childPos = 1;
+        } else {
+          if children[1] != nil {
+            if children[1]!.rank > children[0]!.rank {
+              childPos = 1;
+            }
+          }
+        }
+
+
+        // Rotate the root down
+        // Note that _rotate will change the value of node
+        _rotate(node, childPos);
+
+        // Remove the old root recursively
+        var result = _remove(node!.children[childPos^1], x);
+        node!.update();
+        return result;
+      }
+      else {
+        var pos: int = cmp > 0;
+        var result = _remove(node!.children[pos], x);
+        node!.update();
+        return result;
+      }
+    }
+
+    /*
+      Attempt to remove the item from this orderedSet with a value equal to `x`. If
+      an element equal to `x` was removed from this orderedSet, return `true`, else
+      return `false` if no such value was found.
+
+      :arg x: The element to remove.
+      :return: Whether or not an element equal to `x` was removed.
+      :rtype: `bool`
+    */
+    proc ref remove(const ref x: eltType): bool {
+      var result = false;
+
+      on this {
+        _enter();
+        result = _remove(_root, x);
+        _leave();
+      }
+
+      return result;
+    }
+
+    /*
+      Clear the contents of this orderedSet.
+
+      .. warning::
+
+        Clearing the contents of this orderedSet will invalidate all existing
+        references to the elements contained in this orderedSet.
+    */
+    proc ref clear() {
+      on this {
+        _enter();
+        if _root != nil {
+          delete _root;
+          _root = nil;
+        }
+        _leave();
+      }
+    }
```

Comment:
You are just deleting the root. Don't you think all the child nodes should be deleted?

Reply:
I think this should be handled by _root's deinit